purchases/src/main/kotlin/com/revenuecat/purchases/common/offlineentitlements/ProductEntitlementMapping.kt

```diff
@@ -1,19 +1,29 @@
 package com.revenuecat.purchases.common.offlineentitlements
 
+import com.revenuecat.purchases.common.HTTPResponseOriginalSource
+import com.revenuecat.purchases.common.networking.HTTPResult
+import com.revenuecat.purchases.common.originalDataSource
 import com.revenuecat.purchases.utils.optNullableString
 import org.json.JSONArray
 import org.json.JSONObject
 
 internal data class ProductEntitlementMapping(
     val mappings: Map<String, Mapping>,
+    internal val originalSource: HTTPResponseOriginalSource = HTTPResponseOriginalSource.MAIN,
+    val loadedFromCache: Boolean = false,
 ) {
     companion object {
         private const val PRODUCT_ENTITLEMENT_MAPPING_KEY = "product_entitlement_mapping"
         private const val PRODUCT_ID_KEY = "product_identifier"
         private const val BASE_PLAN_ID_KEY = "base_plan_id"
         private const val ENTITLEMENTS_KEY = "entitlements"
+        private const val ORIGINAL_SOURCE_KEY = "rc_original_source"
 
-        fun fromJson(json: JSONObject): ProductEntitlementMapping {
+        fun fromJson(
+            json: JSONObject,
+            originalSource: HTTPResponseOriginalSource = HTTPResponseOriginalSource.MAIN,
+            loadedFromCache: Boolean = false,
+        ): ProductEntitlementMapping {
```

Comment:
I'm a bit confused here because `fromJson` requires passing `originalSource` when the `toJson()` below does put the `originalSource` into the JSON. Is it correct? or are we losing the `originalSource` stored in the JSON?

Reply:
Yeah, it's indeed confusing... Maybe it would have been worth to refactor it to store an object containing both the product entitlement mapping response + source instead, but wanted to avoid dealing with the migration... For now, I've refactored this a bit [here](https://github.com/RevenueCat/purchases-android/pull/2802/commits/b1ba302dad2cf45c85c2d3e1b26900f54cb27da3), though I noticed I had to create a copy of the json to avoid possible unexpected side effects. I believe it shouldn't be a performance hit the way we're doing it now.